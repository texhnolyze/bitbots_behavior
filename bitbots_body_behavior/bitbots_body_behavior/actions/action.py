from typing import Optional

from bitbots_blackboard.blackboard import BodyBlackboard
from bitbots_body_behavior.state.needs import Needs
from bitbots_body_behavior.state.state import State


class Action:
    def __init__(self, needs: Needs):
        self.needs: list[Needs] = []

    def execute(self, blackboard: BodyBlackboard, new_state: Optional[State]):
        raise NotImplementedError

    def evaluate(self, state: State, new_state: Optional[State]) -> float:
        raise NotImplementedError

    def next_states_to_evaluate(self, state: State) -> list[State]:
        return []

    def __repr__(self):
        return self.__class__.__name__
