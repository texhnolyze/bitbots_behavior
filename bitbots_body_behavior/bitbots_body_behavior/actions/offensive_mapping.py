from bitbots_body_behavior.state.state import State

class OffensiveMapping():
    def __init__(self) -> None:
        pass

    @staticmethod
    def apply(self, role: [int,float]) -> float:
        #Schaut man auf get_role im blackboard, werden diese als Tuple gespeichert, 
        #wobei der int fÃ¼r die entsprechende Rolle steht
        if (role[0] == 3):
            return 0.8
        elif (role[0] == 5):
            return 0.3
        elif (role[0] == 6):
            return 0.1
        else:
            return 0.0
