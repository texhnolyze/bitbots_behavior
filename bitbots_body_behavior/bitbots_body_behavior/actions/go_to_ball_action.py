from bitbots_body_behavior.functions.combinators import AndCombinator, NaturalLogarithm, OrCombinator, Prioritization
from bitbots_body_behavior.functions.utility_functions import (
    EulerExponentialUF,
    ExponentialUF,
    LinearUF,
    NormVerteilungUF,
    PiecewiseUF,
)
from bitbots_body_behavior.state.needs import Need, Needs
from bitbots_body_behavior.state.state import State

from .action import Action


class GoToBallAction(Action):
    def __init__(self, needs: Needs):
        self.needs: list[Need] = [needs.ABLE_TO_MOVE]

    def evaluate(self, state: State) -> float:
        # Block1 der Rolle und allgemeinen Ballposition
        offensive_mapping = PiecewiseUF.setup(LinearUF.setup(0, 0), 0.5, 0.5)
        ball_position = ExponentialUF.setup(1, 1, 1)
        combinator_eins = NaturalLogarithm.apply([offensive_mapping, ball_position], 5)

        # Block2 Winkel und Distanz
        ball_angle = NormVerteilungUF.setup(0.25)
        ball_distance = EulerExponentialUF.setup(1, -1, 1)
        combinator_zwei = AndCombinator.apply(
            [ball_angle.apply(self, state.angle_to_ball), ball_distance.apply(self, state.distance_to_ball)]
        )

        combinator_b1_b2 = Prioritization.apply([combinator_eins, combinator_zwei], [2, 8])

        # Block3 Spielsituation
        goal_difference = PiecewiseUF.setup(LinearUF.setup(-1, 8, 0.5), 4, -4)
        seconds_remaining = PiecewiseUF.setup(LinearUF.setup(-1, 30, 1), 30, 0)
        combinator_drei = Prioritization.apply([goal_difference, seconds_remaining], [9, 1])

        return OrCombinator.apply([combinator_b1_b2, combinator_drei])

    def next_states_to_evaluate(self, state: State) -> list[State]:
        return []
