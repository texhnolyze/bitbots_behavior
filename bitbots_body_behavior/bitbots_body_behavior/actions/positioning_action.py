import math
from typing import Tuple

import numpy as np
from bitbots_msgs.msg import HeadMode
from bitbots_utils.transforms import quat_from_yaw
from geometry_msgs.msg import PoseStamped

from bitbots_blackboard.blackboard import BodyBlackboard
from bitbots_body_behavior.considerations.defensiveness import Defensiveness
from bitbots_body_behavior.considerations.offensiveness import Offensiveness
from bitbots_body_behavior.functions.combinators import (
    OrCombinator,
)
from bitbots_body_behavior.state.needs import Need, Needs
from bitbots_body_behavior.state.state import State

from .action import Action

Point = Tuple[float, float, float]


class PositioningAction(Action):
    def __init__(self, needs: Needs):
        self.needs: list[Need] = [needs.ABLE_TO_MOVE]

    def evaluate(self, _: State, new_state: State) -> float:
        offensiveness = Offensiveness.get_utility_value(new_state)
        defensiveness = Defensiveness.get_utility_value(new_state)

        return OrCombinator.apply([offensiveness, defensiveness])

    def next_states_to_evaluate(self, state: State) -> list[State]:
        # generate the next possible positions in a circle around the current position
        # with an offset of 0.5m in each direction 0°, 45°, 90°, 135°, 180°, 225°, 270°, 315°
        walk_distance = 0.5
        potential_future_positions = np.array(self.generate_potential_positions(state.current_position, walk_distance))
        teammate_positions = np.array(state.active_teammate_poses)

        if len(teammate_positions) > 0:
            # calculate the distance to each teammate resulting in a matrix
            # where each column is a teammate and each row is a potential future position
            distances_to_teammates = np.linalg.norm(
                potential_future_positions[:, np.newaxis, :2] - teammate_positions[np.newaxis, :, :2], axis=2
            )
            # by take the minimum of each row we get the distance to the closest teammate
            # for each potential future position
            distance_to_next_teammate = np.min(distances_to_teammates, axis=1)

            next_positions = potential_future_positions[distance_to_next_teammate > walk_distance]
        else:
            next_positions = potential_future_positions

        return list(map(state.set_current_position, next_positions))

    def generate_potential_positions(self, current_position: Point, distance: float = 1.0) -> list[Point]:
        points = [current_position]
        x, y, theta = current_position

        # setup up the angles for the 8 points around the current position based on the unit circle
        angles = [0, math.pi / 4, math.pi / 2, 3 * math.pi / 4, math.pi, -3 * math.pi / 4, -math.pi / 2, -math.pi / 4]

        for angle in angles:
            new_x = x + distance * math.cos(theta + angle)
            new_y = y + distance * math.sin(theta + angle)
            new_theta = theta + angle

            points.append((new_x, new_y, new_theta))

        return points

    def execute(self, blackboard: BodyBlackboard, new_state: State):
        blackboard.misc.set_head_mode(HeadMode.BALL_MODE)

        pose_msg = PoseStamped()
        pose_msg.header.stamp = blackboard.node.get_clock().now().to_msg()
        pose_msg.header.frame_id = blackboard.map_frame

        pose_msg.pose.position.x = new_state.current_position[0]
        pose_msg.pose.position.y = new_state.current_position[1]
        pose_msg.pose.position.z = 0.0
        pose_msg.pose.orientation = quat_from_yaw(math.radians(new_state.current_position[2]))
        blackboard.pathfinding.publish(pose_msg)
