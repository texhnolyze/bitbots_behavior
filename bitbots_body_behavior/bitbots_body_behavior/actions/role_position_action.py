from bitbots_msgs.msg import HeadMode
from tf2_geometry_msgs import PoseStamped

from bitbots_blackboard.blackboard import BodyBlackboard
from bitbots_body_behavior.state.needs import Need, Needs
from bitbots_body_behavior.state.state import State

from .action import Action


class RolePositionAction(Action):
    def __init__(self, needs: Needs):
        self.needs: list[Need] = [needs.READY_STATE]
        self.ready_state = needs.READY_STATE.available

    def evaluate(self, state: State) -> float:
        if self.ready_state:
            return 1.0
        else:
            return 0.0

    def role_positions(self, blackboard: BodyBlackboard):
        role_positions = blackboard.config["role_positions"]
        kickoff_type = "active" if blackboard.gamestate.has_kickoff() else "passive"
        try:
            if blackboard.team_data.role == "goalie":
                generalized_role_position = role_positions[blackboard.team_data.role]
            else:
                # players other than the goalie have multiple possible positions
                generalized_role_position = role_positions[blackboard.team_data.role][kickoff_type][
                    str(blackboard.misc.position_number)
                ]
        except KeyError as e:
            raise KeyError(f"Role position for {blackboard.team_data.role} not specified in config") from e

        # Adapt position to field size
        # TODO know where map frame is located
        role_position = [
            generalized_role_position[0] * blackboard.world_model.field_length / 2,
            generalized_role_position[1] * blackboard.world_model.field_width / 2,
        ]

        return role_position

    def execute(self, blackboard: BodyBlackboard, state: State):
        blackboard.misc.set_head_mode(HeadMode.BALL_MODE)

        pose_msg = PoseStamped()
        pose_msg.header.stamp = blackboard.node.get_clock().now().to_msg()
        pose_msg.header.frame_id = blackboard.map_frame

        role_position = self.role_positions(blackboard)

        pose_msg.pose.position.x = role_position[0]
        pose_msg.pose.position.y = role_position[1]
        pose_msg.pose.orientation.w = 1.0

        blackboard.pathfinding.publish(pose_msg)
