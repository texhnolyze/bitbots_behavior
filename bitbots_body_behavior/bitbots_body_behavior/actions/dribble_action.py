from typing import Optional

import numpy as np

from geometry_msgs.msg import Twist
from bitbots_blackboard.blackboard import BodyBlackboard
from bitbots_body_behavior.considerations.offensive_mapping import OffensiveMapping
from bitbots_body_behavior.functions.combinators import (
    Prioritization,
)
from bitbots_body_behavior.functions.utility_functions import (
    LinearUF,
    PiecewiseUF,
)
from bitbots_body_behavior.state.needs import Need, Needs
from bitbots_body_behavior.state.state import State

from .action import Action


class DribbleAction(Action):
    def __init__(self, needs: Needs):
        self.needs: list[Need] = [needs.ABLE_TO_MOVE, needs.HAS_BALL]
       
        

    def evaluate(self, state: State) -> float:
        return 1.0

    def next_states_to_evaluate(self, state: State) -> list[State]:
        return []
    
    def execute(self, blackboard: BodyBlackboard, new_state: State | None):
        max_speed_x = blackboard.config["dribble_max_speed_x"]
        min_speed_x = -0.1
        max_speed_y = blackboard.config["dribble_max_speed_y"]
        ball_heading_x_vel_zero_point = blackboard.config["dribble_ball_heading_x_vel_zero_point"]
        p = blackboard.config["dribble_p"]
        max_accel_x = blackboard.config["dribble_accel_x"]
        max_accel_y = blackboard.config["dribble_accel_y"]

        current_speed_x = blackboard.pathfinding.current_cmd_vel.linear.x
        current_speed_y = blackboard.pathfinding.current_cmd_vel.linear.y
       
        ball_v = blackboard.world_model.get_ball_position_uv()
        # Get the relative angle from us to the ball
        ball_angle = blackboard.world_model.get_ball_angle()

        # todo compute yaw speed based on how we are aligned to the goal

        adaptive_acceleration_x = 1 - (abs(ball_angle) / ball_heading_x_vel_zero_point)
        max_speed_x * adaptive_acceleration_x

        current_speed_x = max_accel_x *max_speed_x + current_speed_x * (1 -max_accel_x)

        # give more speed in y direction based on ball position
        y_speed = ball_v * p

        current_speed_y = max_accel_y * np.clip(
            y_speed, -max_speed_y,max_speed_y
        ) + current_speed_y * (1 - max_accel_y)

        cmd_vel = Twist()
        cmd_vel.linear.x = current_speed_x
        cmd_vel.linear.y = current_speed_y
        cmd_vel.angular.z = 0.0
        blackboard.pathfinding.direct_cmd_vel_pub.publish(cmd_vel)

