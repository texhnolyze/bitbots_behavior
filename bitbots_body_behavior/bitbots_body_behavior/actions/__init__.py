# ruff: noqa: F401
from .action import Action
from .dribble_action import DribbleAction
from .go_to_ball_action import GoToBallAction
from .positioning_action import PositioningAction
from .role_position_action import RolePositionAction
from .stand_action import StandAction
