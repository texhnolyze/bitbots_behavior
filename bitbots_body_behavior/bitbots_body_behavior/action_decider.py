import functools
import itertools
from typing import Iterable, Optional, Tuple

from rclpy.impl.rcutils_logger import RcutilsLogger as Logger

from bitbots_blackboard.blackboard import BodyBlackboard

from .actions import Action, DribbleAction, GoToBallAction, PositioningAction
from .execution.process_pool import ProcessPoolExecutor
from .state.needs import Needs
from .state.state import State

Evaluation = Tuple[Action, list[State]]
EvaluationResult = Tuple[Action, State, float]


class Evaluator:
    def __init__(self, state: State):
        self.state = state

    def evaluate(self, evaluation: Evaluation) -> EvaluationResult:
        results = self.evaluate_new_states(evaluation)
        return max(results, key=lambda item: item[2])

    def evaluate_new_states(self, evaluation: Evaluation) -> Iterable[EvaluationResult]:
        action, new_states = evaluation
        return map(
            lambda state: (action, state, action.evaluate(self.state, state)),
            new_states,
        )


class ActionDecider:
    def __init__(self, blackboard: BodyBlackboard, logger: Logger):
        self.state: State
        self.evaluator: Evaluator
        self.next_action: EvaluationResult
        self.blackboard = blackboard

        self.needs = Needs(self.blackboard)
        self.fulfilled_needs = []
        self.actions: list[Action] = self.setup_actions()
        self.next_action: Optional[EvaluationResult] = None

        self.logger = logger
        self.max_parallel_states = 4
        # max_workers = int(len(ALL_ACTIONS) * self.max_parallel_states / 2)
        self.executor = ProcessPoolExecutor(4)

    def setup_actions(self) -> list[Action]:
        return [
            GoToBallAction(self.needs),
            DribbleAction(self.needs),
            PositioningAction(self.needs),
        ]

    def update_state(self):
        self.state.update()
        self.evaluator = Evaluator(self.state)
        self.fulfilled_needs = self.needs.available()

    def decide(self):
        def flat_map(f, args):
            return functools.reduce(lambda a, b: itertools.chain(a, b), map(f, args))

        self.update_state()

        possible_actions = filter(lambda a: self.fulfilled_needs <= a.needs, self.actions)
        needed_evaluations = flat_map(self.split_into_evaluation_parts, possible_actions)
        split_results = self.parallel_evaluation(needed_evaluations)

        if len(list(split_results)):
            ideal_action = max(split_results, key=lambda item: item[2])
            self.next_action = ideal_action

            self.logger.info(f"Ideal action: {ideal_action[0]}, max_score: {ideal_action[2]}, {ideal_action[1]})")

    def execute_ideal_action(self):
        if self.next_action:
            self.logger.info(f"Executing ideal action: {self.next_action[0]}")
            # self.next_action[0].execute(self.next_action[2])

    def parallel_evaluation(self, evaluations: Iterable[Evaluation]) -> list[EvaluationResult]:
        return self.executor.map(self.evaluator.evaluate, evaluations)

    def synchronous_evaluation(self, evaluations: Iterable[Evaluation]) -> list[EvaluationResult]:
        split_results: list[EvaluationResult] = []

        for action, states in evaluations:
            split_results.append(self.evaluator.evaluate((action, states)))

        return split_results

    def split_into_evaluation_parts(self, action: Action) -> Iterable[Evaluation]:
        new_states = action.next_states_to_evaluate(self.state)
        parts = [
            new_states[i : i + self.max_parallel_states] for i in range(0, len(new_states), self.max_parallel_states)
        ]
        return map(lambda part: (action, part), parts)
