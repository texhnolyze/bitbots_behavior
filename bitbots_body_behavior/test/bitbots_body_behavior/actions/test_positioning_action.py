from bitbots_body_behavior.actions import PositioningAction


def test_evalutate(state, new_state, needs):
    action = PositioningAction(needs)
    utility = action.evaluate(state, new_state)

    assert isinstance(utility, float)
    assert utility >= 0.0


def test_next_states_to_evaluate_no_teammates_in_walking_distance(snapshot, needs, state):
    state.current_position = (0, 0, 0)
    state.active_teammate_poses = [(1, 1, 0), (-1, -1, 0)]
    state.set_current_position = lambda x: x

    action = PositioningAction(needs)
    next_states = action.next_states_to_evaluate(state)

    assert len(next_states) == 9
    assert next_states == snapshot


def test_next_states_to_evaluate_teammates_in_walking_distance(snapshot, needs, state):
    state.current_position = (0, 0, 0)
    state.active_teammate_poses = [(0.3, 0.3, 0), (-0.3, 0.3, 0)]
    state.set_current_position = lambda x: x

    action = PositioningAction(needs)
    next_states = action.next_states_to_evaluate(state)

    print(next_states)
    assert len(next_states) == 3
    assert next_states == snapshot
