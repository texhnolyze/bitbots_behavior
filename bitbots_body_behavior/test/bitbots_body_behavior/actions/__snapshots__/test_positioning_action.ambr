# serializer version: 1
# name: test_next_states_to_evaluate_no_teammates_in_walking_distance
  list([
    array([0., 0., 0.]),
    array([0.5, 0. , 0. ]),
    array([0.35355339, 0.35355339, 0.78539816]),
    array([3.06161700e-17, 5.00000000e-01, 1.57079633e+00]),
    array([-0.35355339,  0.35355339,  2.35619449]),
    array([-5.00000000e-01,  6.12323400e-17,  3.14159265e+00]),
    array([-0.35355339, -0.35355339, -2.35619449]),
    array([ 3.06161700e-17, -5.00000000e-01, -1.57079633e+00]),
    array([ 0.35355339, -0.35355339, -0.78539816]),
  ])
# ---
# name: test_next_states_to_evaluate_teammates_in_walking_distance
  list([
    array([-0.35355339, -0.35355339, -2.35619449]),
    array([ 3.06161700e-17, -5.00000000e-01, -1.57079633e+00]),
    array([ 0.35355339, -0.35355339, -0.78539816]),
  ])
# ---
